### ISSUE TITLE
Colour search

### ISSUE THREAD

**Issue by pozieuto** [hydrusnetwork/hydrus#268]
Being able to search or match by colour seems obviously useful. The difficulty here is that images can have very many or very few colours, so it needs to be determined how much information would need to be stored to be useful without becoming infeasible to search through.

With some method for grabbing a given number of dominant colours from an image (I found a performant Python module [here](https://github.com/wenmin-wu/dominant-colors-py)), a reasonable number of dominant colours per image could maybe be determined by incrementally retrieving more and more colours until you get a set where two or more colours are not considered to be "different" enough within some (possibly configurable) HSL/HSV tolerances (or maybe some other perceptual metrics) for it to be worthwhile to keep all of them, or a hard maximum of colours is reached.

Search and duplicate/alternate match functionality could be enhanced with these features:
* Relevancy score: Store the ratio of pixels in the image that each colour best matches to compare against. For matching, enable selection of relevancy tolerance for colours to match.
* Similarity: Provide an interface for selecting a variety of colours by specifying multiple colours and HSL/HSV ranges. For matching, enable selection of HSL/HSV tolerances for colours to match.
* Colour count: Enable searching by the number of dominant colours. For matching, also enable searching by the number of colour matches.

Some useful presets or favourites could be nice here, like just being able to select "mostly" "yellow". Some kind of even more advanced search functionality would also be useful, like being able to search for images with about twice as much red as green.

**Comment by bbappserver** [comment]
Using a colour histogram ultimately comes down to how you want to slice up the spectrum, and is often of limited usefulness for very large datasets.  I would recommend you simply assign an appropriate tag containing a colour name.

**Comment by pozieuto** [comment]
Assigning tags for colours is a prohibitively large amount of manual labour compared to how much time occasional use of searching by colour might save, and would not be nearly as specific. This sort of categorization should be automatable.

**Comment by ShadowJonathan** [comment]
I want to note that its also possible to automate this through a client API, obviously, but the throughput needed would not be beneficial, especially with the fact you'd have to store distinct values (`green`, `limegreen`, `#005348`) instead of being able to sort and search by relative color boundaries/distances, you can do the latter one within hydrus itself, but i'd suggest to make it an optional module that has to be turned on and watching a specific repo or a specific set of tags before it'd be triggered to inspect an image for those colors, to avoid unnecessary load to those who wouldn't use it.

**Comment by pozieuto** [comment]
I agree that this should be opt-in, since it's relatively intensive on both computation and storage.

**Comment by DonaldTsang** [comment]
Color-based image hashing might be useful, however, how would you modify pHash, or other hash for that matter, to do it?
Assuming you want something like color histogram matching, what would be the color resolution (number of hues), and how would you deal with brightness or contrast differences? Would you aggregate the whole image into one single stat, or...
"vague recollection" based search, where you MSPaint a memory of an image to search the original (I know there are software that does this but I forgot where) and that you can remember that certain colors belong in certain parts of an image?

**Comment by pozieuto** [comment]
This proposal as it is does not include any positional information, so it is not like pHashes at all in that way; it only proposes storing dominant colours in an image as a literal list of colour values and ratios. It is limited to searches like ">50% red". (Well, that's a generalization—you'd also have to define what "red" is in terms of hue, saturation, and lightness tolerances.)

However, you _could_ extend this to include very rough positional information in relatively simple ways. For instance, you could also have an 8x8 array storing which of the dominant colours is most present in that section of the image, and you could expose an interface to search against an 8x8 "doodle" using this data. Searching in this way would not be nearly as performant as with actual hashes, though.
